// zenn-embed-elements/lib/classes/katex.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmbedKatex = void 0;
const load_script_1 = require("../utils/load-script");
const load_stylesheet_1 = require("../utils/load-stylesheet");
const containerId = 'katex-container';
class EmbedKatex extends HTMLElement {
    constructor() {
        super();
        const container = document.createElement('div');
        container.setAttribute('id', containerId);
        this._container = container;
    }
    connectedCallback() {
        return __awaiter(this, void 0, void 0, function* () {
            this.render();
        });
    }
    render() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (typeof katex === 'undefined') {
                yield load_script_1.loadScript({
                    src: 'https://cdn.jsdelivr.net/npm/katex@0.12.0/dist/katex.js',
                    id: 'katex-js',
                });
            }
            // CSSを読み込む（まだ読み込まれていない場合のみ読み込まれる）
            load_stylesheet_1.loadStylesheet({
                href: 'https://cdn.jsdelivr.net/npm/katex@0.12.0/dist/katex.min.css',
                id: `katex-css`,
            });
            const displayMode = !!this.getAttribute('display-mode');
            katex === null || katex === void 0 ? void 0 : katex.render((_a = this.innerText) !== null && _a !== void 0 ? _a : '', this._container, {
                macros: { '\\RR': '\\mathbb{R}' },
                throwOnError: false,
                displayMode,
            });
            this.innerHTML = this._container.innerHTML;
        });
    }
}
exports.EmbedKatex = EmbedKatex;
